src/math/sqrt.rs

```diff
@@ -1,3 +1,54 @@
+/// [Square root](https://en.wikipedia.org/wiki/Square_root) of an `f64`.
+///
+/// This function is intended to exactly match the
+/// [`sqrt`](https://en.cppreference.com/w/c/numeric/math/sqrt) function as
+/// defined by the C/C++ spec.
+#[allow(unreachable_code)]
+pub fn sqrt(x: f64) -> f64 {
+    // On most targets LLVM will issue a hardware sqrt instruction instead of a
+    // call to this function when you call `core::intrinsics::sqrtf64(x)`.
+    // However, not all targets have hardware sqrt, and also people might end up
+    // calling this function directly, so we must still do our best to get a
+    // hardware instruction used, and then go with a software fallback if
+    // necessary.
+
+    // Nightly: if we're *sure* that LLVM supports a hardware instruction then
+    // we'll call the LLVM intrinsic. We have to be conservative in our
+    // selection for when to do this, because if the intrinsic usage ends up
+    // calling back here it's infinite recursion.
+    #[cfg(all(
+        feature = "unstable",
+        any(
+            all(target_arch = "x86", not(target_feature = "soft-float")),
```

Comment:
Unfortunately `-C target-feature`, what `--print target-features` does, is not equivalent to `#[cfg(target_feature = "..")]`. I'd recommend looking at compiled code to see how the `#[cfg]` here isn't working.

Reply:
Alright, I see now that `target_feature = "soft-float"` will always evaluate to false (for example), but that brings us back to the question: how do we respect the hard-float/soft-float configuration of the target? If we can't do that, we can't ever be sure that a hardware instruction exists to be called.

Reply:
Sorry I don't always have the answers to all the questions, I'm just pointing out how this isn't working as intended. I don't know how this would be detected.

Reply:
It's totally fine if you don't know it all! You just happen to usually know the most, and you're "here already". We'll have to ask around I guess.

Without an answer to this problem we're just plain stuck. It would be incorrect to call back to the LLVM intrinsic if the arch's version of hardware floating point isn't enabled, since LLVM will end up calling us again and it's just infinite recursion.

The only real relief to be had is that _if there is hardware sqrt_ then LLVM won't ever call us here, so we could just skip trying to divert back to LLVM at all. Except that's where we started basically and that's what _could potentially_ give the performance regressions that we're trying to avoid.

Reply:
Opened an issue to seek a wider audience for support: https://github.com/rust-lang/rust/issues/64514

Reply:
So if I'm understanding the issue it's that if we export `sqrt` and `sqrtf` from `compiler_builtins` unconditionally then any C code that's linked will use this version and it might perform slower than what it's already using. I think the solution is to not export `sqrt` and `sqrtf` unconditionally not only to avoid the performance issue but also because C code could depend on the specific error handling of the libm it was compiled against that this code can't emulate. `core` already depends on `fmod` so I think `sqrt` can and should be handled in the same way.